src/allocator_defrag.c

```diff
@@ -0,0 +1,484 @@
+        }
+    }
+    return info;
+}
+
+/* -----------------------------------------------------------------------------
+ * Interface functions to get fragmentation info from jemalloc
+ * -------------------------------------------------------------------------- */
+#define ARENA_TO_QUERY MALLCTL_ARENAS_ALL
+/**
+ * Initializes the defragmentation system for the jemalloc memory allocator.
+ *
+ * This function performs the necessary setup and initialization steps for the defragmentation system.
+ * It retrieves the configuration information for the jemalloc arenas and bins, and initializes the usage
+ * statistics data structure.
+ *
+ * return 0 on success, or a non-zero error code on failure.
+ *
+ * The initialization process involves the following steps:
+ * 1. Check if defragmentation is supported by the current jemalloc version.
+ * 2. Retrieve the arena bin configuration information using the `je_mallctlbymib` function.
+ * 3. Initialize the `usage_latest` structure with the bin usage statistics and configuration data.
+ * 4. Set the `defrag_supported` flag to indicate that defragmentation is enabled.
+ *
+ * Note: This function must be called before using any other defragmentation-related functionality.
+ * It should be called during the initialization phase of the code that uses the
+ * defragmentation feature.
+ */
+int allocatorDefragInit(void) {
+    if (defrag_supported) return 0;
+    uint64_t epoch = 1;
+    size_t sz = sizeof(epoch);
+    je_mallctl("epoch", &epoch, &sz, &epoch, sz);
+    char buf[100];
+    jeBinInfo *binfo;
+
+    size_t len = sizeof(jemalloc_quantum);
+    je_mallctl("arenas.quantum", &jemalloc_quantum, &len, NULL, 0);
+    // lg-quantum can be 3 or 4
+    assert((jemalloc_quantum == 8) || (jemalloc_quantum == 16));
+
+    unsigned nbins;
+    sz = sizeof(nbins);
+    assert(!je_mallctl("arenas.nbins", &nbins, &sz, NULL, 0));
+    arena_bin_conf.bin_info = zcalloc(sizeof(jeBinInfo) * nbins);
+    for (unsigned j = 0; j < nbins; j++) {
+        binfo = &arena_bin_conf.bin_info[j];
+        /* The size of the current bin */
+        snprintf(buf, sizeof(buf), "arenas.bin.%d.size", j);
+        sz = sizeof(size_t);
+        assert(!je_mallctl(buf, &binfo->reg_size, &sz, NULL, 0));
+
+        /* Number of regions per slab */
+        snprintf(buf, sizeof(buf), "arenas.bin.%d.nregs", j);
+        sz = sizeof(uint32_t);
+        assert(!je_mallctl(buf, &binfo->nregs, &sz, NULL, 0));
+        binfo->len = binfo->reg_size * binfo->nregs;
+        /* Mib of fetch number of used regions in the bin */
+        snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.curregs", j);
+        sz = sizeof(size_t);
+        binfo->info_keys.curr_regs_key_len = sizeof(binfo->info_keys.curr_regs_key) / sizeof(size_t);
+        assert(!je_mallctlnametomib(buf, binfo->info_keys.curr_regs_key, &binfo->info_keys.curr_regs_key_len));
```

Comment:
Definitely a nitpick and feel free to ignore, but this code just feels "crowded" to me. 
```suggestion
        /* Number of regions per slab */
        snprintf(buf, sizeof(buf), "arenas.bin.%d.nregs", j);
        sz = sizeof(uint32_t);
        assert(!je_mallctl(buf, &binfo->nregs, &sz, NULL, 0));
        binfo->len = binfo->reg_size * binfo->nregs;

        /* Mib of fetch number of used regions in the bin */
        snprintf(buf, sizeof(buf), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.curregs", j);
        sz = sizeof(size_t);
        binfo->info_keys.curr_regs_key_len = sizeof(binfo->info_keys.curr_regs_key) / sizeof(size_t);
        assert(!je_mallctlnametomib(buf, binfo->info_keys.curr_regs_key, &binfo->info_keys.curr_regs_key_len));
```